bundles/org.eclipse.ui.workbench/Eclipse UI/org/eclipse/ui/activities/ActivityManagerEvent.java

```diff
@@ -214,4 +302,34 @@ public boolean haveDefinedCategoryIdsChanged() {
 	public boolean haveEnabledActivityIdsChanged() {
 		return enabledActivityIdsChanged;
 	}
+
+	/**
+	 * Returns whether or not enabledActivityIds property changed and any of the
+	 * changed IDs belongs to a non-expression-controlled activity.
+	 *
+	 * @return <code>true>/code> iff the enabledActivityIds property changed and any
+	 *         of the changed IDs belongs to a non-expression-controlled activity.
+	 * @since 3.131
+	 */
+	public boolean haveEnabledNonExpressionControlledActivityIdsChanged() {
+		if (enabledNonExpressionControlledActivityIdsChanged == null) {
+			Set<String> nonNullPreviouslyEnabledActivityIds = previouslyEnabledActivityIds == null ? emptySet()
+					: previouslyEnabledActivityIds;
+			enabledNonExpressionControlledActivityIdsChanged = haveEnabledNonExpressionControlledActivityIdsChanged(
+					activityManager, nonNullPreviouslyEnabledActivityIds);
+		}
+		return enabledNonExpressionControlledActivityIdsChanged;
+	}
+
+	private static boolean haveEnabledNonExpressionControlledActivityIdsChanged(IActivityManager activityManager,
+			Set<String> previousIds) {
+		Set<String> currentIds = activityManager.getEnabledActivityIds();
+		Set<String> unchangedIds = new HashSet<>(currentIds);
+		unchangedIds.retainAll(previousIds);
+		Set<String> changedIds = new HashSet<>(previousIds);
+		changedIds.addAll(currentIds);
+		changedIds.removeAll(unchangedIds);
+		return changedIds.stream().anyMatch(id -> activityManager.getActivity(id).getExpression() == null);
```

Comment:
You could avoid creating new `Set`s if you do it like this:

```java
Set<String> currentIds = activityManager.getEnabledActivityIds();

Stream.concat(currentIds.stream(), previousIds.stream()) // the sum...
		.filter(id -> !currentIds.contains(id) || !previousIds.contains(id))// ...minus the intersection
		.anyMatch(id -> activityManager.getActivity(id).getExpression() == null);
```

You could even experiment concatenating 2 `.parallelStream()`s and see if it improves/degrades performance.

Reply:
Thanks! Good point. It should be even more efficient to treat the streams individually, as then only one `contains` checks for every element is necessary:
```java
	Set<String> currentIds = activityManager.getEnabledActivityIds();
	Stream<String> addedIds = currentIds.stream().filter(id -> !previousIds.contains(id));
	if (addedIds.anyMatch(this::isIdOfNonExpressionControlledActivity)) {
		return true;
	}
	Stream<String> removedIds = previousIds.stream().filter(id -> !currentIds.contains(id));
	return removedIds.anyMatch(this::isIdOfNonExpressionControlledActivity);

...

private boolean isIdOfNonExpressionControlledActivity(String id) {
	return activityManager.getActivity(id).getExpression() == null;
}
```
Actually, this method only exists for compatibility reasons. The actual goal is that the field `enabledNonExpressionControlledActivityIdsChanged` is set via the constructor call as the calculation has implicitly been done before. Thus, in the demo scenario changes to this method have no performance impact. Even if changing the code to explicitly performing the calculation everytime the value is accessed has no measurable performance impact, as the performance issue addressed by this PR does not come from this calculation but from the code that is executed when not performing this check at all.

I have updated the functionality in b8c6a1e accordingly.